### ISSUE TITLE
Support shadow roots as insertion points

### ISSUE THREAD

**Issue by bedeoverend** [cssinjs/jss#539]
Right now setting `insertionPoint` in `options` to a Shadow Root results in:

`Warning: [JSS] Insertion point is not in the DOM.`

See https://www.webpackbin.com/bins/-KpFbaOYG0kxqHAob_lP for replication (open the actual browser console and you should see it logged)

I think it's [this line](https://github.com/cssinjs/jss/blob/master/src/renderers/DomRenderer.js#L181) as a Shadow Root doesn't have a parent node.

IMO it'd be great to have the Shadow Root as an option for an `insertionPoint`, and just append it / prepend it in the top level of that shadow root?

**Comment by kof** [comment]
Huh, didn't know that an element inside of shadow root has no parent.

**Comment by kof** [comment]
Ah you are using shadow root itself as an insertion point. Yeah you could use the first element inside. But yeah, its not optimal.

**Comment by kof** [comment]
Using `element` option is fine though if you have just one style element. If we have real use case for having multiple style elements within one shadow root, we can add a `container` option, which is by default the head of the document and optionally can be a shadow root.

**Comment by kof** [comment]
`shadowRoot.isConnected` seems to be the way to check if it is in the DOM tree or not.

**Comment by bedeoverend** [comment]
Thanks - yeah a `container` option sounds good 👍 

But, as you say, using `element` works fine and I can't see a use case for having multiple stylesheets in the Shadow Root, but this is a relatively unexplored area so far anyway.

**Comment by kof** [comment]
I think it would be interesting to see the performance overhead caused by shadow root.

**Comment by bedeoverend** [comment]
Do you mean just attaching Shadow Roots in general? Or the performance of stylesheets inside Shadow Roots? If I can think of the right setup, I'd be happy to do some benchmarking.

**Comment by kof** [comment]
Both, shadowRoot activation and also overhead when generating a sheet per element with shadowRoot vs reusing same sheet over global styles.

**Comment by kof** [comment]
As of right now, I think this issue is all about an experiment and has no production application, so I am not even sure it is worth fixing.

**Comment by bedeoverend** [comment]
That's fair enough - you're right, for me this is purely experimentation, and I agree I'm not sure if there's a use case for JSS in Shadow Roots.

Understand if you'd like to close this issue, I can come back and comment on it again when I think there's a production use case for JSS + Shadow DOM, and also with performance metrics / considerations.

**Comment by trusktr** [comment]
> no production application, so I am not even sure it is worth fixing

@kof I ran into this, for production application.

I am using `react-shadow` to wrap auto-generated SVGs (f.e. exported from illustrator or somewhere) in ShadowDOM roots for style encapsulation.

Otherwise the class names that these programs generate can cause name clashes, which destroy SVG rendering by cross-styling them with each other's `<style>`s.

But now that they are in Shadow roots, I can not use JSS-generated class names from the outside. :(

---

Basically, I want to wrap SVGs in shadow roots because it is easy to avoid name clashes that way.

Then I want to be able to further style them from the outside.

**Comment by trusktr** [comment]
I only have one JSS stylesheet normally, and I want to apply it to all roots. The `insertionPoint` seems to map one sheet to one insertion point, so I'd have to create many JSS style sheets, one for each `insertionPoint`, which would be a pain if I'm using react-jss, as the JSS sheets are abstracted away and I'm using the decorator/HoC utility. So it's just not straight forward to "use insertionPoints", which would require jumping to hoops.

I think what I have to do is use `insertRule` on all of my shadow dom CSSStyleSheets, to copy from the main one.

**Comment by kof** [comment]
There are 2 options, one is implicit - rendering the same css string in every root and hoping the browser will be able to optimize it internally. Rumors say chrome is doing it.

Another is using the new api for constructable style sheet, which I haven't tried out, nor it is widely supported.

**Comment by trusktr** [comment]
Support for Constructible Style Sheets landed in Chrome 73, and there are polyfills for it now too (the polyfills basically copy the style elements to the shadow roots, like the above idea, but triggered through use of the `adoptedStyleSheets` API).

Can we re-visit this and think about how to make it easy to use JSS with `adoptedStyleSheets`?

**Comment by Paitum** [comment]
I am developing a React microfrontend that is exposed as a Web Component. It uses a ShadowRoot for isolation. Our initial implementation loaded the CSS bundle in the Shadow DOM, which worked, but meant styling was asynchronous. We are now attempting to embed the CSS into the JS payload for immediate availability AND appropriate CSS bundling.

Our POC is using React-JSS, but the styles are being added to the HEAD and don't pierce the Shadow DOM. I am attempting to find a way to have JSS insert the styles in the ShadowRoot. I am not familiar with the JSS life cycle, but it seems JSS wants to create the <style> ASAP, where in my use-case I would expect the Web Component constructor or connectedCallback to be the appropriate moment to ask JSS to create & inject the styles into the DOM within itself.

**Comment by imWildCat** [comment]
Just sharing my finding today. I have been looking for a light-weight solution for this issue for several months. At the end, <https://cssinjs.org/react-jss?v=v10.9.0#server-side-rendering> looks a pretty neat solution for shadow DOMs.

By combining `SheetsRegistry` with `JssProvider`, jss can render just inside the Shadow Root.
